core/lib/pubs.db.test.ts

```diff
@@ -0,0 +1,524 @@
+	CoreSchemaType,
+	ElementType,
+	InputComponent,
+	MemberRole,
+	MembershipType,
+} from "db/public";
+
+import { mockServerCode } from "~/lib/__tests__/utils";
+import { createSeed } from "~/prisma/seed/createSeed";
+
+const { createForEachMockedTransaction } = await mockServerCode();
+
+const { getTrx } = createForEachMockedTransaction();
+
+const seed = createSeed({
+	community: {
+		name: "test",
+		slug: "test-form-view",
+	},
+	users: {
+		admin: {
+			role: MemberRole.admin,
+		},
+		editor: {
+			role: MemberRole.editor,
+		},
+		contributor: {
+			role: MemberRole.contributor,
+		},
+	},
+	pubFields: {
+		Title: { schemaName: CoreSchemaType.String },
+		Description: { schemaName: CoreSchemaType.String },
+		"Some relation": { schemaName: CoreSchemaType.String, relation: true },
+	},
+	pubTypes: {
+		"Basic Pub": {
+			Title: { isTitle: true },
+			"Some relation": { isTitle: false },
+		},
+	},
+	pubs: [
+		{
+			pubType: "Basic Pub",
+			values: {
+				Title: "Some title",
+				Description: "Some description",
+			},
+			members: { admin: MemberRole.admin },
+		},
+		{
+			pubType: "Basic Pub",
+			values: {
+				Title: "Another title",
+			},
+			relatedPubs: {
+				"Some relation": [
+					{
+						value: "test relation value",
+						pub: {
+							pubType: "Basic Pub",
+							values: {
+								Title: "A pub related to another Pub",
+							},
+						},
+					},
+					{
+						value: "second relation value",
+						pub: {
+							pubType: "Basic Pub",
+							values: {
+								Title: "Another pub related to another Pub",
+							},
+						},
+					},
+				],
+			},
+		},
+	],
+	forms: {
+		"basic-pub-form": {
+			pubType: "Basic Pub",
+			elements: [
+				{
+					type: ElementType.pubfield,
+					field: "Description",
+					component: InputComponent.textInput,
+					config: {
+						label: "Description",
+					},
+				},
+				{
+					type: ElementType.pubfield,
+					field: "Title",
+					component: InputComponent.textInput,
+					config: {
+						label: "Title",
+					},
+				},
+			],
+		},
+		"second-form": {
+			pubType: "Basic Pub",
+			elements: [
+				{
+					type: ElementType.pubfield,
+					field: "Title",
+					component: InputComponent.textInput,
+					config: {
+						label: "Title",
+					},
+				},
+				{
+					type: ElementType.pubfield,
+					field: "Some relation",
+					component: InputComponent.relationBlock,
+					config: {
+						relationshipConfig: {
+							label: "Related",
+							help: "Help",
+							component: InputComponent.textInput,
+						},
+					},
+				},
+			],
+		},
+	},
+});
+
+describe("getPubByForm", () => {
+	it("should be able to get a pub with form elements", async () => {
+		const { seedCommunity } = await import("~/prisma/seed/seedCommunity");
+		const { pubs, community, pubTypes } = await seedCommunity(seed);
+		const { getPubByForm } = await import("./pubs");
+		const { getPubsWithRelatedValues } = await import("./server/pub");
+		const { getForm } = await import("./server/form");
+
```

Comment:
sweet, thanks for all the tests!

for the future: you could make all these imports a bit nicer while retaining the auto rollback by creating a small helper function that does the imports, eg

```ts
const getAllImports = async () => {
    	const { seedCommunity } = await import("~/prisma/seed/seedCommunity");
		const { pubs, community, pubTypes } = await seedCommunity(seed);
		const { getPubByForm } = await import("./pubs");
		const { getPubsWithRelatedValues } = await import("./server/pub");
		const { getForm } = await import("./server/form");

        return { seedCommunity, /*...*/ }
}